Dockerfile

```diff
@@ -1,45 +1,81 @@
-FROM ruby:2.5.3-alpine
+FROM ruby:2.5.3-alpine as base
 
 ENV APP_ROOT /usr/src/app
 WORKDIR $APP_ROOT
 
-# =============================================
-# System layer
-
-# Will invalidate cache as soon as the Gemfile changes
-COPY Gemfile Gemfile.lock $APP_ROOT/
-
-# * Setup system
-# * Install Ruby dependencies
-RUN apk add --update \
-    build-base \
+RUN apk add --no-cache \
     netcat-openbsd \
     git \
     nodejs \
     postgresql-dev \
     mysql-dev \
     tzdata \
-    curl-dev \
- && rm -rf /var/cache/apk/* \
+    curl-dev
+
+# Will invalidate cache as soon a the Gemfile changes
+COPY Gemfile Gemfile.lock $APP_ROOT/
+
+# * Setup system
+# * Install common Ruby dependencies
+RUN apk add --no-cache build-base \
  && bundle config --global frozen 1 \
- && bundle install --without test --jobs 2 \
- && gem install foreman
+ && bundle install \
+    --without development test production \
+    --jobs $(grep -c ^processor /proc/cpuinfo) \
+ && gem install foreman \
+ && apk del build-base
 
-# ========================================================
-# Application layer
+# Startup
+CMD ["bin/docker-start"]
+
+####################
+
+FROM base as development
+
+# * Install development dependencies
+RUN apk add --no-cache build-base \
+ && bundle install \
+    --with development \
+    --without test production \
+    --jobs $(grep -c ^processor /proc/cpuinfo) \
+ && apk del build-base
 
 # Copy application code
 COPY . $APP_ROOT
 
-# Precompile assets for a production environment.
-# This is done to include assets in production images on Dockerhub.
-RUN RAILS_ENV=production bundle exec rake assets:precompile
-
 # * Generate the docs
 # * Make files OpenShift conformant
 RUN RAILS_ENV=development bin/rails api_docs:generate \
- && chgrp -R 0 $APP_ROOT \
+ && chown -R 1000:0 $APP_ROOT \
  && chmod -R g=u $APP_ROOT
 
-# Startup
-CMD ["bin/docker-start"]
+USER 1000
+
+####################
+
+FROM base as production
+
+# * Install production dependencies
+RUN apk add --no-cache build-base \
+ && bundle install \
+    --with production \
+    --without development test \
+    --jobs $(grep -c ^processor /proc/cpuinfo) \
+ && apk del build-base
+
+# Copy application code
+COPY . $APP_ROOT/
+
+# Precompile assets for a production environment.
+# This is done to include assets in production images on Dockerhub.
+RUN RAILS_ENV=production rake assets:precompile \
+ && rm -rf /usr/src/app/tmp/cache/
+
+RUN RAILS_ENV=production rake assets:precompile \
```

Comment:
Looks like this line might have been duplicated in the rebase?

Reply:
Oh indeed, that was a rebase mistake. It's now fixed.